pkg/noderesourcetopology/match.go

```diff
@@ -0,0 +1,358 @@
+		resourceBitmask := bm.NewEmptyBitMask()
+		for _, numaNode := range nodes {
+			numaQuantity, ok := numaNode.Resources[resource]
+			// if can't find requested resource on the node - skip (don't set it as available NUMA node)
+			// if unfound resource has 0 quantity probably this numa node can be considered
+			if !ok && quantity.Cmp(zeroQuantity) != 0 {
+				continue
+			}
+			// Check for the following:
+			// 1. set numa node as possible node if resource is memory or Hugepages (until memory manager will not be merged and
+			// memory will not be provided in CRD
+			// 2. set numa node as possible node if resource is cpu and it's not guaranteed QoS, since cpu will flow
+			// 3. set numa node as possible node if zero quantity for non existing resource was requested (TODO check topology manaager behaviour)
+			// 4. otherwise check amount of resources
+			if resource == v1.ResourceMemory ||
+				strings.HasPrefix(string(resource), string(v1.ResourceHugePagesPrefix)) ||
+				resource == v1.ResourceCPU && qos != v1.PodQOSGuaranteed ||
+				quantity.Cmp(zeroQuantity) == 0 ||
+				numaQuantity.Cmp(quantity) >= 0 {
+				resourceBitmask.Add(numaNode.NUMAID)
+			}
+		}
+		bitmask.And(resourceBitmask)
+	}
+}
+
+func (ph PodLevelResourceHandler) PolicyFilter(pod *v1.Pod, zones topologyv1alpha1.ZoneList) *framework.Status {
+	containers := []v1.Container{}
+	containers = append(pod.Spec.InitContainers, pod.Spec.Containers...)
+
+	resources := make(v1.ResourceList)
+
+	for _, container := range containers {
+		for resource, quantity := range container.Resources.Requests {
+			if quan, ok := resources[resource]; ok {
+				quantity.Add(quan)
+			}
+			resources[resource] = quantity
+		}
+	}
+
+	nodes := createNUMANodeList(zones)
+	bitmask := bm.NewEmptyBitMask()
+	bitmask.Fill()
+	checkResourcesForNUMANodes(bitmask, nodes, resources, v1qos.GetPodQOS(pod))
+
+	if bitmask.IsEmpty() {
+		// definitly we can't align container, so we can't align a pod
+		return framework.NewStatus(framework.Unschedulable, fmt.Sprintf("Can't align pod: %s", pod.Name))
+	}
+	return nil
+
+}
+
+func createNUMANodeList(zones topologyv1alpha1.ZoneList) NUMANodeList {
+	nodes := make(NUMANodeList, 0)
+	for _, zone := range zones {
+		if zone.Type == "Node" {
+			var numaID int
+			fmt.Sscanf(zone.Name, "node-%d", &numaID)
+			resources := extractResources(zone)
+			nodes = append(nodes, NUMANode{NUMAID: numaID, Resources: resources})
+		}
+	}
+	return nodes
+}
+
+// Filter Now only single-numa-node supported
+func (tm *NodeResourceTopologyMatch) Filter(ctx context.Context, cycleState *framework.CycleState, pod *v1.Pod, nodeInfo *framework.NodeInfo) *framework.Status {
+	if nodeInfo.Node() == nil {
+		return framework.NewStatus(framework.Error, fmt.Sprintf("Node is nil %s", nodeInfo.Node().Name))
+	}
+	if v1qos.GetPodQOS(pod) == v1.PodQOSBestEffort {
+		return nil
+	}
+
+	nodeName := nodeInfo.Node().Name
+
+	topologyPolicies := getTopologyPolicies(tm.nodeTopologies, nodeName)
```

Comment:
If there is no particular reason (e.g., indexing), it's unnecessary to  maintain a local copy of `nodeTopologies`. Instead, inject a xyzInformerFactory().xyzInfomer.Lister() to the plugin instance, and then use xyzLister's methods to fetch the xyz objects when needed. Then, all the onXYZAdd/Update/Delete methods can also be removed.

Reply:
nodeTopologies are populating in Add/Update callback, it's not a poll model, it's push model, when plugin getting informed about node topology state modification (e.g. after new pod get launched)

Reply:
Oh, you may misunderstand `xyzLister` which is providerd by Kubernetes client-go SDK. It actually maintains a local cache for you, and xyzLister's methods fetch the xyz object(s) from the client cache in an O(1) manner. It's not polling objects from API server.

In other words, no matter what state of your CR gets changed, the cache (xyzStore) that backs xyzLister is (almost) up-to-date.

IIUC, we can just rely on the xyzLister, because we just look at the up-to-date state of particular xyz objects upon a scheduling cycle (the Filter() hook) - rather than needs to be notified and proactively trigger the scheduling of a Pod.